File: ADCSPI/read_serial.py
import serial
import time
from serial.tools import list_ports
def read_serial():
    # List available COM ports
    ports = list(list_ports.comports())
    num_ports = len(ports)

    # Ask the user to input the mode
    mode = int(input('Please enter the mode you want to use (0 for idle, 1 for bin, 2 for event-by-event): '))
    modeChar = mode.to_bytes(1, 'little')

    if mode == 1:
        count = 1+4+2+64+3+2-1
    elif mode == 2:
        count = 1+4+2+2+4+3-1
    elif mode == 0:
        count = 1+4+2+64+3

    for idx, port_info in enumerate(ports):
        print(f"{idx+1}. {port_info.device}")

    # Ask the user to input the COM port number
    port_num = int(input('Please enter the number of the COM port you want to use: ')) - 1
    device = ports[port_num].device

    baud_rate = 115200
    s = serial.Serial(device, baud_rate)

    # Clear input buffer
    s.flushInput()

    message = b'S' + modeChar
    s.write(message)

    dataArray = []
    i = 0

    while True:  # repeat the loop
        data = int.from_bytes(s.read(1), 'little')

        while data != mode:
            data = int.from_bytes(s.read(1), 'little')

        while s.in_waiting < count:
            time.sleep(0.001)

        data = s.read(count)
        current_row = []
        current_row.append(data[0])

        seconds = int.from_bytes(data[3:7], 'little')
        milliseconds = int.from_bytes(data[7:9], 'little')
        current_row.extend([seconds, milliseconds])

        if mode == 2:  # Event-by-event mode
            microseconds = int.from_bytes(data[9:11], 'little')
            current_row.append(microseconds)
            for val in data[11:15]:
                current_row.append(val)
        else:  # Bin mode
            for val in data[11:75]:
                current_row.append(val)

        print(', '.join(map(str, current_row)))
        dataArray.append(current_row)
        s.write(message)

    s.close()

# Call the function if this script is run as the main module
if __name__ == "__main__":
    read_serial()

File: ADCSPI/read_serial.m
function read_serial()
% List available COM ports
ports = serialportlist;
num_ports = length(ports);
sample_speed = 'K0';

dataPoints = 7;

dataArray = [];
i = 1;
j = 1;
for k = 1:num_ports
    fprintf('%d. %s\n', k, ports(k))
end

% Ask the user to input the COM port number
port_num = input('Please enter the number of the COM port you want to use: ');
device = ports(port_num);


baud_rate = 115200;
s = serialport(device, baud_rate);

% Clear input buffer
flush(s);

message = ['S',sample_speed]; 
% Write 'S' to the serial port
write(s, message, "uint8");  % 'S' is cast to uint8 because write requires data in byte format
% write(s, sample_speed,"uint8");

% At this point, data should be aligned

while true
% Read 1 byte for command
data = read(s,1,"uint8");
dataArray(i,j) = data;
j = j + 1;

% Read 4 bytes for seconds
data = read(s,4,"uint8");
int32data = uint32(data(1)) + bitshift(uint32(data(2)), 8) ...
           + bitshift(uint32(data(3)), 16) + bitshift(uint32(data(4)), 24);
dataArray(i,j) = int32data;
j = j + 1;


% Read 2 bytes for miliseconds
data = read(s,2,"uint8");
int16data = uint16(data(1)) + bitshift(uint16(data(2)), 8);
dataArray(i,j) = int16data;
j = j + 1;

for index = 1:4
    data = read(s,1,"uint8");
    dataArray(i,j) = data;
    j = j + 1;
end

for l = 1:dataPoints
    fprintf('%d, ',dataArray(i,l));
end
fprintf('\n');


i = i+ 1;
j = 1;

write(s, message, "uint8"); 
% write(s, sample_speed,"uint8");
% pause(1)

end


end